File: aether_messenger/windows/AetherQMessenger/Crypto/SelfTest.cs
using System;
using System.Text;
using AetherQMessenger.Ffi;

namespace AetherQMessenger.Crypto;

/// <summary>
/// Bitta jarayon ichida Alice/Bob identity'larini simulyatsiya qilib to'liq
/// AUTH-AKEM + Compress-KEM ratchet + record layer round-trip'ini tekshiradi —
/// P/Invoke chegarasini tasdiqlash uchun.
/// </summary>
public static class SelfTest
{
    public sealed record Result(bool Success, string Details);

    public static Result Run()
    {
        try
        {
            var aliceEd = AetherCrypto.GenerateEd25519Identity();
            var bobEd = AetherCrypto.GenerateEd25519Identity();
            var bobX = AetherCrypto.GenerateX25519Identity();
            var bobMlkem = AetherCrypto.GenerateMlKem768Identity();
            var bobZ = AetherCrypto.RandomBytes(32);

            var encaps = AetherCrypto.AkemEncapsulate(
                bobMlkem.PublicKey, bobX.PublicKey,
                aliceEd.SecretKey, aliceEd.PublicKey, bobEd.PublicKey);

            var bobSharedSecret = AetherCrypto.AkemDecapsulate(
                bobMlkem.SecretKey, bobX.SecretKey, bobX.PublicKey, bobZ,
                aliceEd.PublicKey, bobEd.PublicKey,
                encaps.CtMlkem, encaps.CtX25519, encaps.Signature);

            if (!encaps.SharedSecret.AsSpan().SequenceEqual(bobSharedSecret))
            {
                return new Result(false, "AKEM shared_secret mos kelmadi");
            }

            var baseSeed = AetherCrypto.DeriveRatchetSeed(encaps.SharedSecret);
            var sessionId = AetherCrypto.DeriveSessionId(encaps.SharedSecret);

            using var sender = new CompressSenderHandle(sessionId, baseSeed);
            using var recipient = new CompressRecipientHandle(sessionId, baseSeed);

            var step = sender.NextFrame(bobX.PublicKey);
            if (step.Frame.Length != 77)
            {
                return new Result(false, $"Compress frame 77 bayt emas: {step.Frame.Length}");
            }

            var bobStepSecret = recipient.AcceptFrame(step.Frame, bobX.SecretKey);
            if (!step.SharedSecret.AsSpan().SequenceEqual(bobStepSecret))
            {
                return new Result(false, "Ratchet step shared_secret mos kelmadi");
            }

            var plaintext = Encoding.UTF8.GetBytes("AETHER-Q self-test xabari");
            var recordFrame = AetherCrypto.SealRecord(step.SharedSecret, sessionId, plaintext);
            var decrypted = AetherCrypto.OpenRecord(bobStepSecret, sessionId, recordFrame);

            if (!decrypted.AsSpan().SequenceEqual(plaintext))
            {
                return new Result(false, "Record layer roundtrip mos kelmadi");
            }

            return new Result(
                true,
                "AUTH-AKEM + Compress-KEM ratchet + Record layer — barchasi P/Invoke orqali muvaffaqiyatli o'tdi.");
        }
        catch (Exception e)
        {
            return new Result(false, $"Xato: {e.Message}");
        }
    }
}

File: aether_messenger/android/app/src/main/java/com/aetherq/messenger/crypto/SelfTest.kt
package com.aetherq.messenger.crypto

/**
 * Bitta qurilma ichida Alice/Bob identity'larini simulyatsiya qilib to'liq
 * AUTH-AKEM + Compress-KEM ratchet + record layer round-trip'ini tekshiradi.
 * Ikkinchi fizik qurilma/hisob yo'qligi sababli JNI chegarasini shu tarzda tasdiqlaymiz.
 */
object SelfTest {
    data class Result(val success: Boolean, val details: String)

    fun run(): Result =
        try {
            val aliceEd = AetherCrypto.generateEd25519Identity()
            val bobEd = AetherCrypto.generateEd25519Identity()
            val bobX = AetherCrypto.generateX25519Identity()
            val bobMlkem = AetherCrypto.generateMlKem768Identity()
            val bobZ = AetherCrypto.randomBytes(32)

            val encaps =
                AetherCrypto.akemEncapsulate(
                    pkRecipMlkem = bobMlkem.publicKey,
                    pkRecipX25519 = bobX.publicKey,
                    skSenderSign = aliceEd.secretKey,
                    pkSenderSign = aliceEd.publicKey,
                    pkRecipSign = bobEd.publicKey,
                )
            val bobSharedSecret =
                AetherCrypto.akemDecapsulate(
                    skRecipMlkem = bobMlkem.secretKey,
                    skRecipX25519 = bobX.secretKey,
                    pkRecipX25519 = bobX.publicKey,
                    zRecip = bobZ,
                    pkSenderSign = aliceEd.publicKey,
                    pkRecipSign = bobEd.publicKey,
                    ctMlkem = encaps.ctMlkem,
                    ctX25519 = encaps.ctX25519,
                    signature = encaps.signature,
                )
            check(encaps.sharedSecret.contentEquals(bobSharedSecret)) { "AKEM shared_secret mos kelmadi" }

            val baseSeed = AetherCrypto.deriveRatchetSeed(encaps.sharedSecret)
            val sessionId = AetherCrypto.deriveSessionId(encaps.sharedSecret)
            val aliceSender = AetherCrypto.newCompressSender(sessionId, baseSeed)
            val bobRecipient = AetherCrypto.newCompressRecipient(sessionId, baseSeed)

            val step = aliceSender.nextFrame(bobX.publicKey)
            check(step.frame.size == 77) { "Compress frame 77 bayt emas: ${step.frame.size}" }
            val bobStepSecret = bobRecipient.acceptFrame(step.frame, bobX.secretKey)
            check(step.sharedSecret.contentEquals(bobStepSecret)) { "Ratchet step shared_secret mos kelmadi" }

            val plaintext = "AETHER-Q self-test xabari".toByteArray(Charsets.UTF_8)
            val recordFrame = AetherCrypto.sealRecord(step.sharedSecret, sessionId, plaintext)
            val decrypted = AetherCrypto.openRecord(bobStepSecret, sessionId, recordFrame)
            check(decrypted.contentEquals(plaintext)) { "Record layer roundtrip mos kelmadi" }

            Result(true, "AUTH-AKEM + Compress-KEM ratchet + Record layer — barchasi JNI orqali muvaffaqiyatli o'tdi.")
        } catch (e: Exception) {
            Result(false, "Xato: ${e.message}")
        }
}
